src/datachain/data_storage/metastore.py

```diff
@@ -1864,28 +1912,36 @@ def _checkpoints_query(self):
             *[getattr(self._checkpoints.c, f) for f in self._checkpoints_fields]
         )
 
-    def create_checkpoint(
+    def get_or_create_checkpoint(
         self,
         job_id: str,
         _hash: str,
         partial: bool = False,
         conn: Any | None = None,
     ) -> Checkpoint:
         """
-        Creates a new job query step.
+        Creates a new checkpoint or returns existing one if already exists.
+        This is idempotent - calling it multiple times with the same job_id and hash
+        will not create duplicates.
         """
-        checkpoint_id = str(uuid4())
-        self.db.execute(
-            self._checkpoints_insert().values(
-                id=checkpoint_id,
-                job_id=job_id,
-                hash=_hash,
-                partial=partial,
-                created_at=datetime.now(timezone.utc),
-            ),
-            conn=conn,
+        # First check if checkpoint already exists
+        query = self._checkpoints_insert().values(
+            id=str(uuid4()),
+            job_id=job_id,
+            hash=_hash,
+            partial=partial,
+            created_at=datetime.now(timezone.utc),
         )
-        return self.get_checkpoint_by_id(checkpoint_id)
+
+        # Use on_conflict_do_nothing to handle race conditions
+        if hasattr(query, "on_conflict_do_nothing"):
```

Comment:
why do we have this protective hasattr here? 

even if we have it - it must be assert then